### PR TITLE
switch laravel mix to vite

### LINKED ISSUE #395: switch to vite
we use laravel mix to build our javascript, which is odd since this isn't a laravel project - and even laravel is using vite now so it's time for us to make the switch

the important thing that mix does is bundle our javascript into a single, predictable url at `/public/app.js`. we also need to confirm that the minified size is roughly the same: ~1.4m

then update the [steps for contributing](https://github.com/code4recovery/tsml-ui?tab=readme-ov-file#contributing) in readme

### PR DESCRIPTION
**Pull request by joshreisner**
creating this as a draft PR of three issues, and i don't know enough yet about cjs vs es modules to solve them. the issues are:

1. shipping this would require us to add `type="module"` to everyone's app scripts

2. also `npm run jest` doesn't run because of the error:

```
ReferenceError: module is not defined in ES module scope
This file is being treated as an ES module because it has a '.js' file extension and 'tsml-ui/package.json' contains "type": "module". To treat it as a CommonJS script, rename it to use the '.cjs' file extension.
```

3. finally, it's creating an extra app script for mapbox-gl in `dist/assets/` which is not ideal -- not sure what to do about that

### CODE REVIEW

File: example.html [hunk 1]
```diff
@@ -53,6 +53,6 @@
           : {},
       };
     </script>
-    <script src="/app.js" async></script>
+    <script src="/src/app.tsx" type="module" async></script>
```
Comment (joshreisner) [inline example.html:56]: this is just the example - the actual embed codes haven't changed

File: .eslintrc.js [hunk 2]
```diff
@@ -10,11 +10,6 @@ module.exports = {
     'plugin:import/errors',
     'plugin:import/typescript',
   ],
-  globals: {
-    Lang: 'readonly',
-    TSMLReactConfig: 'readonly',
-    Translation: 'readonly',
-  },
```
Comment (joshreisner) [inline .eslintrc.js:17]: i could not get jest to recognize the global types defined in `index.d.ts` so they're no longer global - also `TSMLReactConfig` got renamed to `Settings`

File: package.json [hunk 3]
```diff
@@ -3,61 +3,51 @@
...
-    "@code4recovery/spec": "1.0.9",
-    "@emotion/react": "^11.11.1",
+    "@code4recovery/spec": "^1.1.4",
+    "@emotion/react": "^11.13.3",
     "deepmerge": "^4.3.1",
-    "luxon": "^3.3.0",
-    "react": "^18.2.0",
-    "react-dom": "^18.2.0",
+    "luxon": "^3.5.0",
+    "mapbox-gl": "^3.7.0",
```
Comment (joshreisner) [inline package.json:45]: it complained that this was missing - we were referencing files in it - so now it's an explicit dependency

File: public/data/aasepia.json
Comment (joshreisner) [inline public/data/aasepia.json:1]: i removed all these test files in favor of a single `example.html` - i think this is more vite-y

File: src/helpers/filter-meeting-data.ts [hunk 4]
```diff
@@ -33,15 +33,11 @@ export function filterMeetingData(
...
-            state.input[filter].map(
-              key => getIndexByKey(state.indexes[filter], key)?.slugs ?? []
-            )
-          )
-        );
+        matchGroups.push([
+          ...state.input[filter]
+            .map(key => getIndexByKey(state.indexes[filter], key)?.slugs ?? [])
+            .flat(),
+        ]);
```
Comment (joshreisner) [inline src/helpers/filter-meeting-data.ts:40]: had to fix all the `// @ts-expect-error`s for jest - these places are the only places where syntax changes. i tested this of course but if there's a regression it's in one of these places

File: src/helpers/format-search-params.ts
Comment (joshreisner) [inline src/helpers/format-search-params.ts:1]: fun fact we weren't using this file

**Comment by joshreisner** [comment]
i've switched out https://aasanjose.org/meetings to use the new app bundle

File: vite.config.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+      input: {
+        main: resolve(__dirname, 'example.html'),
+      },
+      output: {
+        entryFileNames: 'app.js',
+        manualChunks: undefined,
+        inlineDynamicImports: true,
+      },
+    },
+    target: 'es2021',
```
Comment (gkovats) [inline vite.config.ts:19]: @joshreisner  when I add here:
`copyPublicDir: false,`
It doesn't copy logo.svg and a few pointless files in `/dist` folder. Was just poking at various config options.

File: tsconfig.app.tsbuildinfo [hunk 6]
```diff
@@ -0,0 +1 @@
...s","./src/helpers/format-address.ts","./src/helpers/format-array.ts","./src/helpers/format-conference-provider.ts","./src/helpers/format-directions-url.ts","./src/helpers/format-feedback-email.ts","./src/helpers/format-ics.ts","./src/helpers/format-slug.ts","./src/helpers/format-string.ts","./src/helpers/format-url.ts","./src/helpers/get-index-by-key.ts","./src/helpers/index.ts","./src/helpers/load-meeting-data.ts","./src/helpers/query-string.ts","./src/helpers/settings.ts","./src/helpers/translate-google-sheet.ts","./src/helpers/user-agent.ts","./src/i18n/en.ts","./src/i18n/es.ts","./src/i18n/fr.ts","./src/i18n/index.ts","./src/i18n/ja.ts","./src/i18n/nl.ts","./src/i18n/pt.ts","./src/i18n/sk.ts","./src/i18n/sv.ts","./src/styles/alert.ts","./src/styles/button.ts","./src/styles/controls.ts","./src/styles/dropdown.ts","./src/styles/global.ts","./src/styles/index.ts","./src/styles/loading.ts","./src/styles/map.ts","./src/styles/meeting.ts","./src/styles/table.ts","./src/styles/variables.ts","./src/types/jsondata.ts","./src/types/meeting.ts","./src/types/metatype.ts","./src/types/settings.ts","./src/types/state.ts","./src/types/translation.ts","./src/types/index.ts"],"version":"5.6.2"}
```
Comment (gkovats) [inline tsconfig.app.tsbuildinfo:1]: When I build this locally, this file changes - `/src/components/alert.tsx` becomes `/src/components/Alert.tsx`, etc. Case of files match local file paths. Not sure if this is a difference between iOS and Linux.

**Review by gkovats — APPROVED** [review]
@joshreisner Clearly it builds and works. Congrats on getting it to work! I know Vite is the new big build tool, it's opinionated and simplifies stuff. Seems nice, not having to run 2 commands to see locally. I want to update `/example.html` shortly to allow for multiple test meetings, or we'd just change it to test production issues as needed and revert before commit.

Like I mentioned in chat, only things I saw:
- adds 400kb to size, but there's a lot of polyfill voodoo a build tool adds, figure it's in that.
- I don't know why it requires async on my local, but that's a concern if we make this version live. Don't know if you can make every user also update TSML, or folks that have one-off html embeds. Maybe we could offer some helper JS that just embeds the latest script tag necessary for TSML-UI?
- looking at the size, it just prompted me to want to try another map library, just to possibly trim down the size of the widget. I know, bandwidth is cheap, gzip, and caching and all, but some folks are still stuck on 3G, so little things help.

File: public/data/aasepia.json
Comment (gkovats) [reply public/data/aasepia.json:1]: sure?

File: vite.config.ts (hunk 5 above)
Comment (joshreisner) [reply vite.config.ts:19]: thanks - `logo.svg` is needed though for the https://tsml-ui.code4recovery.org/ homepage 👍

File: tsconfig.app.tsbuildinfo (hunk 6 above)
Comment (joshreisner) [reply tsconfig.app.tsbuildinfo:1]: huh, it might be mac vs linux! we can keep an eye on that - i'd kind of prefer this file to use the correct casing

**Comment by joshreisner** [comment]
yeah, shifting over to es modules would probably help too, since that lib file would get cached, i think. 

having our users switch away from mapbox to a different service might be a headache, if there's a way to avoid that. or maybe we might look at other mapbox configurations, maybe we don't need all these dependencies?

i haven't been able to repro this `async` issue but it might make sense to release another version of TSML with that option in it so if users run into problems they can always update

File: vite.config.ts (hunk 5 above)
Comment (gkovats) [reply vite.config.ts:19]: Oh! This is the public page, neat, makes sense.

**Pull request merged**